File: content/snippets/typescript/design_patterns/interpreter.ts
interface Expression {
    interpret(ctx: string): boolean;
}

class TerminalExpression implements Expression {
    constructor(private literal: string) {}
    
    interpret(ctx: string): boolean {
        return ctx.includes(this.literal);
    }
}

class OrExpression implements Expression {
    constructor(private expr1: Expression, private expr2: Expression) {}
    
    interpret(ctx: string): boolean {
        return this.expr1.interpret(ctx) || this.expr2.interpret(ctx);
    }
}

class AndExpression implements Expression {
    constructor(private expr1: Expression, private expr2: Expression) {}
    
    interpret(ctx: string): boolean {
        return this.expr1.interpret(ctx) && this.expr2.interpret(ctx);
    }
}

const c1 = new TerminalExpression("cat");
const c2 = new TerminalExpression("dog");
const expr = new OrExpression(c1, c2);

console.log(expr.interpret("cat"));
console.log(expr.interpret("bird"));

File: content/snippets/go/design_patterns/interpreter.go
package main

import (
	"fmt"
	"strings"
)

type Expression interface {
	Interpret(string) bool
}

type TerminalExpression struct {
	literal string
}

func (t *TerminalExpression) Interpret(ctx string) bool {
	return strings.Contains(ctx, t.literal)
}

type OrExpression struct {
	expr1, expr2 Expression
}

func (o *OrExpression) Interpret(ctx string) bool {
	return o.expr1.Interpret(ctx) || o.expr2.Interpret(ctx)
}

type AndExpression struct {
	expr1, expr2 Expression
}

func (a *AndExpression) Interpret(ctx string) bool {
	return a.expr1.Interpret(ctx) && a.expr2.Interpret(ctx)
}

func main() {
	c1 := &TerminalExpression{"cat"}
	c2 := &TerminalExpression{"dog"}
	expr := &OrExpression{c1, c2}

	fmt.Println(expr.Interpret("cat"))
	fmt.Println(expr.Interpret("bird"))
}
